File: 4_regression/polyreg.py
"""
Created on Sat Mar 18 18:34:06 2017

@author: Robert
"""

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
plt.style.use('seaborn-deep')


ds = pd.read_csv('Position_Salaries.csv')
X = ds.iloc[:,1:2].values
y = ds.iloc[:, -1].values

# Linear regression
from sklearn.linear_model import LinearRegression

lin_reg = LinearRegression()
lin_reg.fit(X,y)

# Polynomial regression
from sklearn.preprocessing import PolynomialFeatures

poly_reg = PolynomialFeatures(3)
X_poly = poly_reg.fit_transform(X)
regressor = LinearRegression()
regressor.fit(X_poly, y)

# So that we can see the complete prediction curve
X_grid = np.arange(min(X), max(X), 0.1)
X_grid = X_grid.reshape(len(X_grid), 1)

plt.scatter(X, y)
plt.plot(X_grid, regressor.predict(poly_reg.fit_transform(X_grid)))
plt.plot(X, lin_reg.predict(X))
plt.xlabel('Position')
plt.ylabel('Salary')
plt.title('Polynomial regression salary progression')
plt.show()

File: 4_regression/polyreg.r
ds = read.csv("Position_Salaries.csv")
ds = ds[, 2:3]

# Fitting and predicting
lin_reg = lm(Salary ~.,
             data = ds)

poly_reg = lm(Salary ~ poly(Level, 4),
              data = ds)
y_pred = predict(poly_reg, data.frame(Level =6.5))

# Plotting
X_grid = seq(min(ds$Level), max(ds$Level), 0.1)

library(ggplot2)
ggplot() +
    geom_point(aes(x= ds$Level, y = ds$Salary), 
               colour = 'red') +
    geom_line(aes(x=ds$Level, y = predict(lin_reg, new_data = X_grid)),
              colour = 'blue') +
    geom_line(aes(x = X_grid, y = predict(poly_reg, newdata = data.frame(Level = X_grid))),
              colour = 'black') +
    ggtitle("Polynomial regression salary progression") +
    xlab("Position") +
    ylab("Salary")
